### ISSUE TITLE
Question re: fine-tuning

### ISSUE THREAD

**Issue by kashyapakshay** [Yangr116/VST#4]
Hi, this looks great and thank you for open-sourcing! I'm trying to fine-tune the 3B SFT model for a custom visual reasoning task (using a robot's gripper cam feeds) and am following the Huggingface [Qwen2 fine-tuning guide](https://huggingface.co/learn/cookbook/en/fine_tuning_vlm_trl) - does that make sense or are there any additional steps needed?

Especially want to make sure these LoRA settings are okay - 
```
peft_config = LoraConfig(
		lora_alpha=16,
		lora_dropout=0.05,
		r=8,
		bias="none",
		target_modules=["q_proj", "v_proj"],
		task_type="CAUSAL_LM",
	)
```

Thanks.

**Comment by kashyapakshay** [comment]
Actually, it's improved quite a bit with more training, no quantization, and using the 7B checkpoint. May have posted this prematurely haha. I'll post my training code and hyperparams anyway, would love to know if you notice any opportunities for further improvement.

```
model_name = "rayruiyang/VST-7B-SFT"
processor = AutoProcessor.from_pretrained(model_name, min_pixels=256*28*28, max_pixels=1280*28*28)
model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
	model_name,
	dtype=torch.bfloat16,
	attn_implementation="flash_attention_2",
).to(DEVICE)

...

peft_config = LoraConfig(
	lora_alpha=16,
	lora_dropout=0.05,
	r=8,
	bias="none",
	target_modules=["q_proj", "v_proj"],
	task_type="CAUSAL_LM",
)

training_args = SFTConfig(
	output_dir="saved/vst-7b-fn",
	num_train_epochs=3000,
	per_device_train_batch_size=4,
	per_device_eval_batch_size=4,
	gradient_accumulation_steps=8,
	gradient_checkpointing_kwargs={"use_reentrant": False},
	max_length=None,
	optim="adamw_torch_fused",
	learning_rate=2e-4,
	# Mixed precision and gradient settings
	bf16=True,
	max_grad_norm=0.3,
	warmup_ratio=0.03,
	# Logging and evaluation
	...
)

trainer = SFTTrainer(
	model=model,
	args=training_args,
	train_dataset=...,
	eval_dataset=...,
	peft_config=peft_config,
	processing_class=processor,
)
trainer.train()
```

<img width="530" height="435" alt="Image" src="https://github.com/user-attachments/assets/8763bdf3-2a37-4f63-992e-628fc676c253" />
<img width="966" height="427" alt="Image" src="https://github.com/user-attachments/assets/edb6bef3-8d4f-4656-b043-bee0526dbe00" />

**Comment by Yangr116** [comment]
Very nice! There is a loss spike at the start of the training, so you can reduce the learning rate.

**Comment by kashyapakshay** [comment]
In case this helps someone in the future - I also figured out an issue where images weren't being loaded/processed right (so the model was basically blind). I'm loading data from a PyTorch dataset into HF's Dataset for SFTTrainer, and needed to implement a custom collate_fn that inserts images and processes them using qwen_vl_utils' `process_vision_info`:

```python
# trainer
from datasets import Dataset as HFDataset

def dataset_generator(dataset):
	for sample in dataset:
		yield sample["data"] # {'images': [...], 'messages': [...]}

trainer = SFTTrainer(
	model=model,
	args=training_args,
	train_dataset=HFDataset.from_generator(dataset_generator, gen_kwargs={'dataset': train_dataset}),
	eval_dataset=HFDataset.from_generator(dataset_generator, gen_kwargs={'dataset': val_dataset}),
	peft_config=peft_config,
	processing_class=processor,
	data_collator=lambda b: vst_collate_fn(b, processor=processor),
)

# collator
def vst_collate_fn(batch, processor):
	examples = [item['messages'] for item in batch]

	# insert images
	images = [item['images'] for item in batch]
	for i in range(len(examples)):
		for j in range(len(examples[i])):
			if examples[i][j]['role'] == 'user':
				for msg in examples[i][j]['content']:
					if msg['type'] == 'image':
						msg['image'] = images[i][msg['index']]
						del msg['text']

	texts = [
		processor.apply_chat_template(example, tokenize=False)
		for example in examples
	]

	image_inputs = [
		process_vision_info(example)[0]
		for example in examples
	]

	model_inputs = processor(
		text=texts,
		images=image_inputs,
		return_tensors="pt",
		padding=True
	)

	labels = model_inputs["input_ids"].clone()
	labels[labels == processor.tokenizer.pad_token_id] = -100
	image_tokens = [151652, 151653, 151655]

	for image_token_id in image_tokens:
		labels[labels == image_token_id] = -100

	model_inputs["labels"] = labels

	return model_inputs
```
